### PR TITLE
[ML-9745] make_spark_converter() converts MLlib vectors into dense arrays

### PR DESCRIPTION
**Pull request by liangz1**
Tests will be added when pyspark 3.0.0 is released.

### CODE REVIEW

File: petastorm/spark/spark_dataset_converter.py [hunk 1]
```diff
@@ -303,10 +303,34 @@ def _convert_precision(df, precision):
...
+    import pyspark
+    if pyspark.__version__ < "3.0.0":
+        raise ValueError("Vector columns are not supported for pyspark<3.0.0.")
+    # pylint: disable=import-error
+    from pyspark.ml.functions import vector_to_array
+    # pylint: enable=import-error
+
+    for struct_field in df.schema:
+        col_name = struct_field.name
+        if struct_field.dataType in {Vectors, SparseVector, OldVector}:
```
Comment (WeichenXu123) [inline petastorm/spark/spark_dataset_converter.py]: Vectors is not the vector base class.

Use `from pyspark.ml.linalg import Vector`
Remove importing SparseVector class (why do you need it here ?)

**Review by WeichenXu123 — CHANGES_REQUESTED** [review]
* Merge master
* Add test (in pytest style)

File: petastorm/spark/spark_dataset_converter.py (hunk 1 above)
Comment (liangz1) [reply petastorm/spark/spark_dataset_converter.py]: Right, we don't support SparseVector now.

**Comment by WeichenXu123** [comment]
Add a test, and check if spark version < 3.0 skip the test (use `pytest.skip` to do this).

Once spark 3.0 released to pypi, we can add a CI conf for spark 3.0. But currently, we can merge this PR first.

File: petastorm/spark/spark_dataset_converter.py [hunk 2]
```diff
@@ -391,10 +391,35 @@ def _gen_cache_dir_name():
...
+    from pyspark.ml.linalg import Vector
+    from pyspark.mllib.linalg import Vector as OldVector
+
+    types_set = {struct_field.dataType for struct_field in df.schema}
+    found_vectors = not types_set.isdisjoint({Vector, OldVector})
+    if not found_vectors:
+        return df
+
+    import pyspark
+    if pyspark.__version__ < "3.0.0":
```
Comment (WeichenXu123) [inline petastorm/spark/spark_dataset_converter.py]: Use:
```
from distutils.version import LooseVersion
LooseVersion(pyspark.__version__) >= '3.0'
```

The same in test.

File: petastorm/tests/test_spark_dataset_converter.py [hunk 3]
```diff
@@ -320,6 +322,29 @@ def test_array(test_ctx):
...
+    converter1 = make_spark_converter(df)
+    with converter1.make_tf_dataset() as dataset:
+        iterator = dataset.make_one_shot_iterator()
+        tensor = iterator.get_next()
+        with tf.Session() as sess:
+            ts = sess.run(tensor)
+    assert np.float32 == ts.vec.dtype.type
+    assert np.float32 == ts.oldVec.dtype.type
+    assert (2, 3) == ts.vec.shape
+    assert (2, 3) == ts.oldVec.shape
```
Comment (WeichenXu123) [inline petastorm/tests/test_spark_dataset_converter.py:346]: Also test compare vector content.

Comment (WeichenXu123) [inline petastorm/tests/test_spark_dataset_converter.py:346]: Also test compare vector content.

**Pull request merged**